[JINA] first PR and add laddergame step1

NORIKIM opened the pull request:
첫 pr과 함께 사다리 게임 1단계 완료하고 커밋,푸시 하였습니다.

NORIKIM committed:
Add Laddergame step1
implement feature
1. get participant by readline
   참가자 입력받기
2. get ladder's height by readline
   사다리 높이 입력 받기
3. make ladder by 2D array
   2차원 배열로 사다리 만들기
4. create random number for add footboard later
   추후 발판 추가를 위한 랜덤숫자 생성
5. add footboard on the ladder then print it
   사다리에 발판을 추가하고 출력

godrm commented on LadderGame/LadderGame/main.swift:
```diff
@@ -2,9 +2,92 @@
+    print("참여할 사람은 몇 명 인가요?")
+    let people = readLine()
+    return Int(people!)!
+}
```
제가 getPeople() 함수 이름으로 유추한 것은 People struct 구조체를 리턴할 것 같습니다. 
함수 이름에 타입을 넣지 않고 리턴값의 구체적인 의미를 명시하는 방법이 좋습니다.
그리고 readLine() 처럼 옵셔널을 리턴하는 경우는 강제로 `!`를 붙여서 unwrapping 하기보다는 nil 일때 처리할 기본값을 주는 방식을 권장합니다. 특히 `Int(people!)!` 구문은 people에 숫자가 아닌 문자가 포함된 경우 항상 프로그램이 멈춰버릴 위험이 있습니다.

godrm commented on LadderGame/LadderGame/main.swift:
```diff
@@ -2,9 +2,92 @@
+    print("최대 사다리 높이는 몇 개인가요?")
+    let ladderHeight = readLine()
+    return Int(ladderHeight!)!
+}
```
getLadderHeight() 에서 LadderHeight는 의미있는 표현이라 좋습니다.
`Int(ladderHeight!)!` 마찬가지로 위험한 코드를 개선해보세요.

godrm commented on LadderGame/LadderGame/main.swift:
```diff
@@ -2,9 +2,92 @@
+// 사다리 만들기 구조체로 만들기...ing
+let people = getPeople()
+let ladderHeight = getLadderHeight()
+let colCount = people + (people - 1)
```
선언한 함수들을 사용하는 호스트 코드는 이렇게 중간에 나오지 않고 맨 아래 모아두세요.
왜냐하면 이후 단계부터는 함수들 선언부와 호출하는 부분을 별도 파일로 분리할 예정입니다. 
그리고 이렇게 쓰면 people이나 ladderHeight 변수는 글로벌 변수가 되서 동일한 이름을 사용할 때 실수를 할 수 있습니다.

godrm commented on LadderGame/LadderGame/main.swift:
```diff
@@ -2,9 +2,92 @@
+            print(resultLadder[raw][col], terminator: "")
+        }
+        print()
+    }
```
printLadder() 함수가 하나 이상의 역할과 책임을 갖고 있는 것 같습니다.
makeFootBoard()를 외부에서 호출해서 사다리 2차원 배열을 만들어서 매개변수로 넘겨주는 구조였으면 좀 더 명확할 것 같습니다.

godrm commented on LadderGame/LadderGame/main.swift:
```diff
@@ -2,9 +2,92 @@
+        madeLadder[raw[i]][col[i]] = "-"
+    }
+    return madeLadder
+}
```
사다리 발판을 만드는 구조는 여러 가지 방식이 있을 수 있습니다. 아래 질문들을 답변해보세요.
- 사다리 정보를 2차원 배열로 해야한다면, String 이 좋을까, Int 가 좋을까
- 지금처럼 2차원 배열을 미리 만들고 채워넣을지, 순차적으로 만드는게 좋을까
- 랜덤값을 어떤 데이터 타입으로 다루는 게 좋을까

NORIKIM replied:
1. 사다리 정보를 2차원 배열로 해야한다면, String 이 좋을까, Int 가 좋을까
사다리가 출력되었을 때의 형태자체가 string이므로 int타입으로 진행 하는 것보다 코드가 더 효율적이라고 생각했습니다.  int타입으로 2차원 배열을 만들게 되면 int에서 string으로 형변환을 해줘야 하고 형변환 시 nil 체크까지 필요하게 될거라 생각했구요.. 
 
2. 지금처럼 2차원 배열을 미리 만들고 채워넣을지, 순차적으로 만드는게 좋을까
2차원배열을 순차적으로 만든다는건 어떤 말씀이신지 잘 이해가 안되요..ㅠㅠ 

3. 랜덤값을 어떤 데이터 타입으로 다루는 게 좋을까
전체적으로 타입 자체를 boolean타입으로 바꾸고 메모리를 더 적게 사용하는 방식으로 바꿔야 좀 더 좋은걸까요?

godrm replied:
개발자가 늘 데이터구조를 선택할 때는 메모리구조가 얼마나 영향을 미치는지 고민해서 선택해야 합니다.
사다리 발판을 표현하는 데이터는 최소한 필요한 데이터만 갖고 있다가 화면에 표시할 때 `-`로 할지, `---`로 할지, 아니면 iOS 화면에 사다리를 그릴지 View객체가 판단할 수도 있습니다.
타입마다 메모리구조는 크기가 다른데, Bool은 1, Int는 8, String은 24+이상을 차지합니다. 최소한의 메모리구조를 선택하는 기준을 항상 고민해보세요.

godrm reviewed (commented):
첫 PR 축하합니다. 아래 피드백을 고민해서 코멘트를 남기거나 개선해보세요.

NORIKIM committed:
Modify code
1. rename getPeople() -> getPlayer()
2. change processing optional unwrapping.
3. the host code move into the while statement on the bottom of the code.

NORIKIM committed:
Modify code
A makeFootBoard() was move into the while statement(main) which calls from printLadder() then pass to printLadder's parameter.

NORIKIM commented:
피드백받은 내용 반영하고 다시 push했습니다. (일단 데이터 구조에 대한 내용은 수정하진 않은 상태입니다.)

godrm commented:
1단계 내용을 충분히 표현한 것 같습니다. 다음 단계에서 데이터 구조도 개선해보세요 👍
